Implemented get_name in StatsLogger, updated Otel and StatsD

ArshiaZr opened the pull request:
- Added get_name method in StatsLogger (base_stat_logger.py) to handle metric name preparation and tag validation.
- Updated Otel and StatsD loggers to inherit from StatsLogger and utilize the new get_name method.
- Refactored tag preparation and validation logic into get_name for a cleaner and more consistent implementation.

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #42881 
related: #42881 

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

ArshiaZr committed:
Implemented get_name in StatsLogger, updated Otel and StatsD

- Added get_name method in StatsLogger (base_stat_logger.py) to handle metric name preparation and tag validation.
- Updated Otel and StatsD loggers to inherit from StatsLogger and utilize the new get_name method.
- Refactored tag preparation and validation logic into get_name for a cleaner and more consistent implementation.

ArshiaZr committed:
Implemented get_name in StatsLogger, updated Otel and StatsD

- Added get_name method in StatsLogger (base_stat_logger.py) to handle metric name preparation and tag validation.
- Updated Otel and StatsD loggers to inherit from StatsLogger and utilize the new get_name method.
- Refactored tag preparation and validation logic into get_name for a cleaner and more consistent implementation.

howardyoo reviewed (approved):
Looks good to me, and ready. Thank you for pulling this up.

ferruzzi commented on airflow/metrics/base_stats_logger.py:
```diff
@@ -31,55 +33,53 @@ class StatsLogger(Protocol):
 
     instance: StatsLogger | NoStatsLogger | None = None
 
-    @classmethod
```
Why are we removing @classmethod from all of these?  Seems unusual to remove the decorator but leave the `cls` parameter.

shahar1 replied:
I'd be happy for an explanation about the necessity of removing `@classmethod` - won't it break the functionality that we have in `dag_processing/manager.py`? For example:

https://github.com/apache/airflow/blob/1c0adb21f6cb5efa24270913444f189776835747/airflow/dag_processing/manager.py#L290C45-L290C46

ferruzzi commented on airflow/metrics/base_stats_logger.py:
```diff
@@ -31,55 +33,53 @@ class StatsLogger(Protocol):
         raise NotImplementedError()
 
+    def get_name(self, metric_name: str, tags: Attributes | None = None) -> str:
+        pass
```
Consider raising `NotImplementedError()` instead of passing, as seen in `timer()` above.  This will force any implementations (StatsD, Datadog, OTel, whatever comes later) to implement it.

ferruzzi commented on airflow/metrics/otel_logger.py:
```diff
@@ -209,22 +210,26 @@ def incr(
-            counter = self.metrics_map.get_counter(full_name(prefix=self.prefix, name=stat), attributes=tags)
+        full_metric_name = self.get_name(metric_name, tags)
+
+        if self.metrics_validator.test(full_metric_name) and name_is_otel_safe(self.prefix, full_metric_name):
```
(Here and below) Is there a reason not to move the `name_is_otel_safe()` check into the otel implementation of `get_name()`?  Why have `get_name` return a name which is not safe to use?  That would also reduce the duplicate code in each of these methods.

ferruzzi commented on airflow/metrics/otel_logger.py:
```diff
@@ -271,24 +277,33 @@ def gauge(
-        if self.metrics_validator.test(stat):
-            self.metrics_map.set_gauge_value(full_name(prefix=self.prefix, name=stat), value, delta, tags)
+        full_metric_name = self.get_name(metric_name, tags)
+        if self.metrics_validator.test(full_metric_name):
```
If you do not move the `name_is_otel_safe()` check into `get_name()` then it looks like it was missing from this check and should be added here to match the other methods.

ferruzzi commented on airflow/metrics/statsd_logger.py:
```diff
@@ -44,27 +46,38 @@
+        if metric_name is None:
+            metric_name = ""
+
+        if self.influxdb_tags_enabled and tags is not None:
```
```suggestion
        if self.influxdb_tags_enabled and tags:
```

ferruzzi commented on airflow/metrics/statsd_logger.py:
```diff
@@ -44,27 +46,38 @@
     @wraps(fn)
     def wrapper(
-        self, stat: str | None = None, *args, tags: dict[str, str] | None = None, **kwargs
+        self, metric_name: str | None = None, tags: dict[str, str | int] | None = None
```
Did you run into a case where this `int` needed to be added?   Leaving this as a `str` doesn't seem like a bad idea at face value.

ferruzzi commented on airflow/metrics/statsd_logger.py:
```diff
@@ -44,27 +46,38 @@
-                            log.error("Dropping invalid tag: %s=%s.", k, v)
-        return fn(self, stat, *args, tags=tags, **kwargs)
+        # if not isinstance(self, SafeStatsdLogger):  # Check if self is an instance of SafeStatsdLogger
+        #     raise TypeError("Expected an instance of SafeStatsdLogger")
```
I think you meant to remove this?

ferruzzi commented on airflow/metrics/statsd_logger.py:
```diff
@@ -44,27 +46,38 @@
+
+        if self.influxdb_tags_enabled and tags is not None:
+            valid_tags: dict[str, str | int] = {}
+            for k, v in tags.items():
```
Rather than casting key and value to strings and storing their new values in separate variables, you could cast the whole dict in place before this loop using a comprehension: 

```
tags = {str(k):str(v) for k, v in tags.items()}
```

then just use k and v directly in this loop since you know they are already strings.

ferruzzi commented on airflow/metrics/statsd_logger.py:
```diff
@@ -44,27 +46,38 @@
+                    if all(c not in [",", "="] for c in value_str) and all(
+                        c not in [",", "="] for c in key_str
+                    ):
+                        valid_tags[k] = value_str
```
If you are going to keep using the case `key_str` value then I think that should likely be used here as well instead of the raw `k`, no?

ferruzzi commented on tests/core/test_otel_logger.py:
```diff
@@ -356,3 +356,110 @@ def test_timer_start_and_stop_manually_send_true(self, mock_time, metrics_consis
         )
+
+    def test_incr_counter(self):
+        # Arrange
```
I see your pattern and appreciate that you were taught this way, and may be required in classes, but in the Airflow codebase the Arrange/Act/Assert comments aren't needed.  A blank line separating the sections of the test is sufficient, "everyone" knows what the intent behind the three "sections" is without adding extra lines of text to the code.

I had not seen AAA before, I kinda like that naming.  I was taught it as "Given, When, Then".  Same concept, just different names.

ferruzzi commented on tests/core/test_otel_logger.py:
```diff
@@ -356,3 +356,110 @@ def test_timer_start_and_stop_manually_send_true(self, mock_time, metrics_consis
+
+    def test_incr_counter(self):
+        # Arrange
+        logger = self.stats
```
Nitpick: the term "logger" is a bit overloaded in Airflow and this could be misread as logger meaning the console like log.error().  Consider renaming this (here and below) to `otel` or `otel_logger` just to help clarify, or just using `self.stats` directly.

ferruzzi commented on tests/core/test_otel_logger.py:
```diff
@@ -356,3 +356,110 @@ def test_timer_start_and_stop_manually_send_true(self, mock_time, metrics_consis
+        # Assert
+        assert isinstance(timer.duration, float)
+        assert timer.duration >= expected_duration - 0.1
+        assert timer.duration <= expected_duration + 0.1
```
Non-blocking Nitpick: `0.1` is a "magic number", consider creating a variable called "acceptable_margin" or something like that.

ferruzzi commented on tests/core/test_otel_logger.py:
```diff
@@ -356,3 +356,110 @@ def test_timer_start_and_stop_manually_send_true(self, mock_time, metrics_consis
         self.meter.get_meter().create_observable_gauge.assert_called_once_with(
             name=full_name(name), callbacks=ANY
         )
+
```
One of the difficulties of unit testing is trying to cover all the edge cases.   All of the tests you added are "happy cases" where things were expected to work and we make sure that they did.  Can you think of any way to add negative tests where we make sure bad things are caught and handled as expected?   For example, what is expected to happen if we try to call `get_name()`  (for example when we create a new counter) with an invalid name?  Should `get_name()` return that invalid name? Should it log a warning?  Something else entirely?

What other ways could we break the new code, and does the code handle those edge cases as expected?

Ideas:

Call `get_name()` directly with:
1.  an invalid name. t
2. try it with and without tags, 
3. etc.  

In theory not including tags on the otel one shouldn't change anything, but having that unit test is a form of regression testing, not just making sure your code works now but ensuring nobody breaks it later on accident.  So this may seem like a silly test to include now, but it can save a lot of trouble down the road.

ferruzzi reviewed (commented):
Left some comments and questions, overall looking pretty good but still needs a little work.

ArshiaZr committed:
Refactor StatsLogger base class and update SafeOtelLogger and SafeStatsdLogger for enhanced validation

- Modified StatsLogger base class to enforce stricter validation on metric names and prefixes, ensuring all names are strings and conform to required formats.
- Enhanced SafeOtelLogger with additional validation to support OpenTelemetry standards and added regression tests for metrics with and without tags.
- Updated SafeStatsdLogger to align with new validation standards and to handle edge cases more robustly.
- Added regression tests to catch issues with missing tags, empty names, and overly long names, ensuring consistency across changes.

ArshiaZr committed:
Refactor StatsLogger base class and update SafeOtelLogger and SafeStatsdLogger for enhanced validation

- Modified StatsLogger base class to enforce stricter validation on metric names and prefixes, ensuring all names are strings and conform to required formats.
- Enhanced SafeOtelLogger with additional validation to support OpenTelemetry standards and added regression tests for metrics with and without tags.
- Updated SafeStatsdLogger to align with new validation standards and to handle edge cases more robustly.
- Added regression tests to catch issues with missing tags, empty names, and overly long names, ensuring consistency across changes.

ArshiaZr committed:
Merge branch 'main' into main

ferruzzi commented on tests/core/test_otel_logger.py:
```diff
@@ -358,108 +358,134 @@ def test_timer_start_and_stop_manually_send_true(self, mock_time, metrics_consis
+        assert timer.duration >= expected_duration - acceptable_margin
+        assert timer.duration <= expected_duration + acceptable_margin
         assert self.meter.get_meter().create_observable_gauge.call_count == 1
         self.meter.get_meter().create_observable_gauge.assert_called_once_with(
```
You can actually drop the `call_count == 1` assertion here since `assert_called_once` also asserts exactly once.  [[docs]](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once_with)

ferruzzi commented on airflow/metrics/otel_logger.py:
```diff
@@ -212,10 +212,8 @@ def incr(
-            counter = self.metrics_map.get_counter(
-                full_name(prefix=self.prefix, name=full_metric_name), attributes=tags
-            )
+        if self.metrics_validator.test(full_metric_name):
```
I see what you tried to do but this ends up being a breaking change.  By moving the `full_name()` call into the `get_name()`, we are now passing the full name (with prefix) into the validator.test() call which will break all user's Allow/Block lists.

If you choose to add the `full_name()` call in there then you must strip `prefix` and `delimiter` off before passing it into `validator.test()`, OR add some logic to test() which does that before checking the name/pattern match.  Those both feel like acceptable options, but maybe feel like more work than leaving the `full_name()` calls where they are.

Which, in hindsight, brings up the naming of the variable `full_metric_name`.  If it includes the prefix, that sounds good, but without the prefix then maybe that needs to be renamed.  Naming is one of the hardest parts of the job.  Perhaps `base_name` since it gets the prefix added later?

ferruzzi commented on airflow/metrics/otel_logger.py:
```diff
@@ -241,8 +239,8 @@ def decr(
 
         full_metric_name = self.get_name(metric_name, tags)
-        if self.metrics_validator.test(full_metric_name) and name_is_otel_safe(self.prefix, full_metric_name):
-            counter = self.metrics_map.get_counter(full_name(prefix=self.prefix, name=full_metric_name))
```
So much cleaner :heart_eyes:   I love it when existing code can be streamlined.

ferruzzi commented on airflow/metrics/statsd_logger.py:
```diff
@@ -51,26 +51,22 @@ def prepare_metric_name_with_tags(fn: T) -> T:
     @wraps(fn)
     def wrapper(
-        self, metric_name: str | None = None, tags: dict[str, str | int] | None = None
+        self, metric_name: str | None = None, tags: dict[str, str] | None = None
```
If you are no longer accepting `int` as a value type here, then I don't think you need to do the casting on L60?   I'm, not sure why you had added `int` as a valid type and then removed it thought, so one or the other should be fixed.

If `tags` is going to be a `dict[str, str]` then there is no need to cast it to a `dict[str, str]` on L60.  If it is going to be declared as `dict[str, str | int]` then we should keep the casing.

ferruzzi replied:
We've gone back and forth on this one a couple times.  Can you confirm that it makes sense how it is now, or did I confuse you and you are just appeasing me?  If you need to chat about it, hit me up on Slack and we can talk it through.

ferruzzi commented on tests/core/test_otel_logger.py:
```diff
@@ -358,108 +358,134 @@ def test_timer_start_and_stop_manually_send_true(self, mock_time, metrics_consis
+            full_name(prefix=self.stats.prefix, name=metric_name), tags
         )
         assert current_value_after_decrement == 0
+
```
Non-blocking nitpick/style suggestion, feel free not to implement this idea;

You could combine all of the following new tests into one parametrized test which passes in the name being tested, the tags (or perhaps a bool indicating whether to use a set of tags which are defined in the test itself), and the expected exception message.   If you'd like to see what that might look like, check out [this](https://github.com/apache/airflow/blob/main/providers/tests/amazon/aws/operators/test_appflow.py#L212) test

ferruzzi reviewed (commented):
Looking much better, left a few more comments.  One big thing that has to be fixed, but the rest are little.

ArshiaZr committed:
Refactor metric name validation to preserve Allow/Block list compatibility

ArshiaZr committed:
Refactor metric name validation to preserve Allow/Block list compatibility

ArshiaZr committed:
Merge branch 'main' of https://github.com/ArshiaZr/airflow

ferruzzi commented on airflow/metrics/otel_logger.py:
```diff
@@ -271,24 +274,29 @@ def gauge(
-        if self.metrics_validator.test(stat) and name_is_otel_safe(self.prefix, stat):
+        if dt is None:
+            log.warning("The duration (dt) cannot be None. Skipping timing update.")
+            return
```
Can you explain this addition for me?

ferruzzi commented on airflow/metrics/otel_logger.py:
```diff
@@ -300,6 +308,18 @@ def timer(
+        if not metric_name:
+            raise InvalidStatsNameException("The stat name cannot be None or an empty string.")
+
+        base_metric_name = metric_name
```
Is there a reason not to use `metric_name` directly now that this method has been cleaned up?  It appears like you are not modifying either version of the parameter, so there shouldn't be a need to make a copy of it.

ferruzzi commented on airflow/metrics/statsd_logger.py:
```diff
@@ -44,27 +46,33 @@
-                            log.error("Dropping invalid tag: %s=%s.", k, v)
-        return fn(self, stat, *args, tags=tags, **kwargs)
+        if metric_name is None:
+            metric_name = ""
```
Purely a style nitpick, your call.  Python's ternary assignment would be handy here:
```suggestion
        metric_name = metric_name or ""
```

ferruzzi commented on airflow/metrics/statsd_logger.py:
```diff
@@ -44,27 +46,33 @@
+
+            for k, v in tags.items():
+                if self.metric_tags_validator.test(k):
+                    if all(c not in [",", "="] for c in v) and all(c not in [",", "="] for c in k):
```
```suggestion
                   if all(c not in [",", "="] for c in f"{v}{k}"):
```
Since we are checking if those symbols are in either k or v, then checking if it is in f'{k}{v}' works just fine and looks much neater.

ferruzzi commented on airflow/metrics/statsd_logger.py:
```diff
@@ -79,80 +87,84 @@ def __init__(
+    def get_name(self, metric_name: str, tags: Attributes | None = None) -> str:
+        """Get metric name with tags, ensuring no invalid keys are included."""
+        if tags:
+            return f"{metric_name},{','.join(f'{k}={v}' for k, v in tags.items())}"
```
Please verify this join.    I seem to remember this should be joined on an underscore, not an equals sign?

ArshiaZr replied:
it is equal sign as it was before I guess

ferruzzi replied:
Aright, maybe I'm wrong. :+1:

ArshiaZr committed:
Merge branch 'main' into main

ArshiaZr committed:
Merge branch 'main' into main

ArshiaZr committed:
Refactored and deleted unnecessary parts of the code

ArshiaZr committed:
Refactored and deleted unnecessary parts of the code

ArshiaZr committed:
Merge branch 'main' into main

ferruzzi commented on airflow/metrics/statsd_logger.py:
```diff
@@ -79,80 +86,84 @@ def __init__(
-        if stat and self.metrics_validator.test(stat):
-            return Timer(self.statsd.timer(stat, *args, **kwargs))
+        full_metric_name = self.get_name(metric_name, tags)
+        if full_metric_name and self.metrics_validator.test(full_metric_name):
```
Why is this the only one that gets `if full_metric_name`?

ArshiaZr replied:
Because the timer is the only one that could accept None as metric_name the way it was before. I can change it so the metric_name can't be None, so we don't check that. I just wanted to make it consistent with the way it was before.

ferruzzi replied:
For now, leave it how you have it I think.   I'm not sure offhand why that was the only one that allowed the option not to provide a name but I'm also still on flu meds and not thinking as clearly as I might. :P  If that is the only one that accepts None for a name, then having that check is good.

ferruzzi commented on tests/core/test_otel_logger.py:
```diff
@@ -356,3 +356,116 @@ def test_timer_start_and_stop_manually_send_true(self, mock_time, metrics_consis
         self.meter.get_meter().create_observable_gauge.assert_called_once_with(
             name=full_name(name), callbacks=ANY
         )
+
```
Taking a closer look at the tests cases, the new  `test_incr_counter` and `test_decr_counter` add testing for a new metric with a count, which is new ground not covered by the existing tests.  Nice catch on those!  

Could you please move those into the same "block" as their cousins so all the incr and decr tests are together, rename them to match the existing convention (ie `test_incr_new_metric_with_count` ), and have them use the `name` fixture like the others?

The gauge and timer ones I'm not sure what is being tested now that wasn't already covered.  What am I missing?

ArshiaZr replied:
That's right actually. So I removed the newly added tests for gauge and timer and moved the test_incr, and test_decr to the right place with proper names.

ferruzzi replied:
Nice catch on those new test cases!

ArshiaZr committed:
Categorized the test cases accordingly and named them properly

ArshiaZr committed:
Categorized the test cases accordingly and named them properly

ArshiaZr committed:
Merge branch 'main' of https://github.com/ArshiaZr/airflow

ferruzzi reviewed (approved):
Nice work.

ashb commented on airflow/metrics/otel_logger.py:
```diff
@@ -300,6 +305,15 @@ def timer(
         """Timer context manager returns the duration and can be cancelled."""
         return _OtelTimer(self, stat, tags)
 
+    def get_name(self, metric_name: str, tags: Attributes | None = None) -> str:
```
Tags doesn't seem to be used. Should it be?

ferruzzi replied:
For Otel, no.    StatsD should append the tags tot he name, OTel should just pass name through.

ferruzzi replied:
The TLDR on this PR is that StatsD emits metrics as something like `local_task_job.task_exit.<job_id>.<dag_id>.<task_id>.<return_code>`  and doesn't support tags but OTel emits it as `local_task_job.task_exit` with a name length limit of (32??) characters but allows a tags dict for the rest, so rather than emitting everything twice, which is done in a bunch of places in the code, this change should allow metrics to be emitted once and whatever metrics backend the user uses can assemble the name (or not) as it needs.  So StatsD will append the tags dict tot he name, OTel will just pass the name through as provided. Whatever future backend may get added can handle it however it needs to


Once this is done and merged, Arshia can go through and remove the duplicates and clean the code up a bit.

ashb replied:
32 char length limit on metric? Is that our choice (in which case it feels easy too short) or otel protocol limitation?

ArshiaZr replied:
Can't they have 255 characters based on this: https://opentelemetry.io/docs/specs/otel/metrics/api/

ferruzzi replied:
I just checked the OTel docs.  ~~I was mistaken on the length, it's a 63-character limit imposed on their end,~~ [edit:  that was for the unit name, not the metric name]  But that goes fast when we are embedding values like dag_id, task_id, operator_name and stuff like that in the name.

ferruzzi replied:
Alright, I wasn't completely hallucinating... https://github.com/open-telemetry/opentelemetry-java/issues/5656

```
Description
OTEL Agent 1.28.0 applies rule "metric name not to exceed 63 characters" and drops metrics (does not push them to Collector) that break this rule.
```

ferruzzi replied:
That is specifically about the Java Agent, but I wonder if it was/is a bug in the underlying otel-collector.   I know I definitely ran into it when I implemented the OTel metrics and that's why we had the different shorter names in the first place, and `name_is_otel_safe` enforces that length.

howardyoo replied:
So, I dug through the otel python source and found the following:
<img width="516" alt="image" src="https://github.com/user-attachments/assets/588c6c49-f459-460b-8ac9-f8cd199c91d7">
So apparently, 255 characters supported.

ArshiaZr replied:
That's what I also found

ferruzzi replied:
Cool, maybe it was a bug they fixed on their end.   Best way to tell is to make a metric with a long name and test it out using `breeze start-airflow --integration otel` and see if it works.

If that limit no loner exists then..... I don't know.... I think we should still use tagging but I don't know what the best approach is for the name itself.  It makes sense to emit the longer name the same as statsd, but then that's another breaking change since existing otel dashboards would all break..... I guess that would need a community discussion at that point.


@ArshiaZr  - Next step, please get a breeze environment running with otel and verify the name limit restriction has been removed/fixed, then we can go from there.

@dannyl1u is working on the metrics registry which you are going to use to auto-generate the docs page, so I guess we need to determine which name we're going to be using before that gets too far....

ashb replied:
> So, I dug through the otel python source and found the following: <img alt="image" width="516" src="https://private-user-images.githubusercontent.com/32691630/382109888-588c6c49-f459-460b-8ac9-f8cd199c91d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA0NzE3ODQsIm5iZiI6MTczMDQ3MTQ4NCwicGF0aCI6Ii8zMjY5MTYzMC8zODIxMDk4ODgtNTg4YzZjNDktZjQ1OS00NjBiLThhYzktZjhjZDE5OWM5MWQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAxVDE0MzEyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM4ZDhmZTYxOWE3MDU0Yzk5OWIwOWJlY2Q4MDYwZGU2ZTk2N2I5N2YwMTUxMmEwOGFmOTMzNjVjMDJjYjllY2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.RkiN_hbqna4PuwWGjckmZoAP_upCjm8QrTvV7dVSb8A"> So apparently, 255 characters supported.

If otel is validating the things, do we need to also validate it, or can we simply rely on it to do the job for us and not have the complexity? (My vote is almost always to have less code)

ferruzzi replied:
I agree if OTel is doing that now and it's working as expected, then we can let it.  I also think that's getting out of scope for this PR though.  I think we can implement this "get name" and worry about testing and possibly pruning out the name validation if they are doing it on their end.

I also agree with Howard that we can and should keep the OTel names short with tags and continue to emit the long concatenated name for StatsD, which means this get_name implementation is still needed/useful either way.

How do you feel about merging this, and Arshia can experiment with removing the name validation we have implemented and make sure the otel validation is actually working as advertised?   I'm the one who initially implemented this in our codebase and I wouldn't have done all that name validation code if their validation was actually working.  I remember getting frustrated trying to figure out the "undocumented" length limit at the time.  If they've fixed it now then by all means let's clean our code up!

howardyoo reviewed (approved):
LGTM

ashb reviewed (approved):
The datadog statsd format natively supports tags -- we should use those there/and it should be updated to inherit form StatsLogger too

ashb reviewed (dismissed):
The datadog statsd format natively supports tags -- we should use those there/and it should be updated to inherit form StatsLogger too

ArshiaZr committed:
Refactor tag handling in DogStatsdLogger

- Added prepare_stat_with_tags decorator to standardize tag processing across all metrics functions.
- Ensured tags default to an empty list [] when no valid tags are provided or self.metrics_tags is disabled, matching expected behavior in tests.
- Updated all metric functions (e.g., incr, decr, gauge, timing, timer) to use the new decorator for consistent tag validation and formatting.

ArshiaZr committed:
Refactor tag handling in DogStatsdLogger

- Added prepare_stat_with_tags decorator to standardize tag processing across all metrics functions.
- Ensured tags default to an empty list [] when no valid tags are provided or self.metrics_tags is disabled, matching expected behavior in tests.
- Updated all metric functions (e.g., incr, decr, gauge, timing, timer) to use the new decorator for consistent tag validation and formatting.

ArshiaZr committed:
Implemented inheritance for SafeDogStatsdLogger

ArshiaZr committed:
Implemented inheritance for SafeDogStatsdLogger

ferruzzi commented on airflow/metrics/datadog_logger.py:
```diff
@@ -51,7 +55,33 @@
     )
 
 
```
Nice work seeing and extracting the common code here, it really helps clean up the whole class.   If you are making this and statsd both inherit from StatsLogger, could you move this decorator into the base class for both to use, or am I missing some difference between them?

ArshiaZr replied:
There are two differences mainly that's why I didn't do that:

1. Tags Format:
prepare_stat_with_tags converts tags into a tags_list (e.g., ["key:value"]) for use in kwargs.
prepare_metric_name_with_tags keeps tags as a dictionary, formatted for InfluxDB.

2. Return Behavior:
prepare_stat_with_tags runs the function directly with tags_list in kwargs.
prepare_metric_name_with_tags returns a modified metric_name with tags if influxdb_tags_enabled is True.

ferruzzi replied:
Sounds good then.

ArshiaZr committed:
Merge branch 'main' into main

ferruzzi committed:
Merge branch 'main' into main

howardyoo reviewed (approved):
Thanks!

ashb commented on airflow/metrics/datadog_logger.py:
```diff
@@ -66,109 +96,95 @@ def __init__(
         rate: float = 1,
         *,
-        tags: dict[str, str] | None = None,
+        tags: list[str] | None = None,
```
I see what's going on here (the decorator changes the dict into the list) but I wonder if there's an easier way of dealing with this... 🤔

ArshiaZr replied:
Since Datadog expecting a list I don't think if we can do better here. Since the decorator would be called first, it does not break any test cases. I can revert it back, but there would be many redundant codes for each of the functions.

ashb replied:
I'm more worried about type completion, but I guess if nothing is complaining this is fine.

ashb commented on airflow/metrics/datadog_logger.py:
```diff
@@ -51,7 +55,33 @@
+            valid_tags: dict[str, str] = {}
+            for k, v in tags.items():
+                if self.metric_tags_validator.test(k):
+                    if all(c not in [",", "="] for c in f"{v}{k}"):
```
Why do we care about `=` being in the key or value? Isn't the separator for datadog tags `:` so this should be
 ```suggestion
                     if ":" not in f"{v}{k}":
```

ashb replied:
Shouldn't this be the job of `self.metric_tags_validator` though?

ArshiaZr replied:
As far as I know, it's not checking it that's why we have the same checks in Statd.

ArshiaZr replied:
But you're right I change the checks for "," and "=" to ":"

ashb commented on airflow/metrics/datadog_logger.py:
```diff
@@ -51,7 +55,33 @@
+
+    @wraps(fn)
+    def wrapper(self, stat: str | None = None, *args, tags: dict[str, str] | None = None, **kwargs):
+        stat = stat or ""
```
Nit: the `stat` has been renamed to `metric_name` elsewhere, so lets do it too.

```suggestion
    def wrapper(self, metric_name: str | None = None, *args, tags: dict[str, str] | None = None, **kwargs):
        metric_name = metric_name or ""
```

ashb commented on airflow/metrics/datadog_logger.py:
```diff
@@ -51,7 +55,33 @@
+
+    @wraps(fn)
+    def wrapper(self, stat: str | None = None, *args, tags: dict[str, str] | None = None, **kwargs):
+        stat = stat or ""
```
Why do we need this `stat or ""`? Shouldn't this decorator pass through whatever it's given?

ArshiaZr replied:
Because in the StatsLogger all of the functions get "str" as metric_name and not "str | None"

ashb replied:
Not _all_ of them though -- timer/timing accepts None doesn't it?

ArshiaZr replied:
Yeah we decided to let only timer/timing accept None. Is it something that need to change

ArshiaZr replied:
Also in the new commit I deleted them both from datalog and statd.

ashb commented on airflow/metrics/statsd_logger.py:
```diff
@@ -44,27 +46,32 @@
-                        else:
-                            log.error("Dropping invalid tag: %s=%s.", k, v)
-        return fn(self, stat, *args, tags=tags, **kwargs)
+        metric_name = metric_name or ""
```
Ditto here? Why aren't we leaving None as None?

ArshiaZr replied:
Same as the previous explanation.

ArshiaZr replied:
You're right. After digging to the code, I realized the none case is being handled by the other decorator (@validate_stat). I changed them both

ashb commented on airflow/metrics/statsd_logger.py:
```diff
@@ -44,27 +46,32 @@
     @wraps(fn)
     def wrapper(
-        self, stat: str | None = None, *args, tags: dict[str, str] | None = None, **kwargs
+        self, metric_name: str | None = None, tags: dict[str, str] | None = None
```
Nit: you changed tags from keyword only to positional with this change.

```suggestion
        self, metric_name: str | None = None, *, tags: dict[str, str] | None = None
```

ashb replied:
This also possibly breaks things as `stats.incr("metric_a", 2). Please bring `*args` back.

ArshiaZr replied:
"prepare_metric_name_with_tags()" is only being called before "get_name()" which is only being called in the functions. Also when I added them back it broke some of the previously written test cases

ashb commented on tests/core/test_stats.py:
```diff
@@ -485,7 +485,7 @@ def test_increment_counter(self):
         self.stats.incr(
             "test_stats_run.delay",
-            tags={"key0": 0, "key1": "val1", "key2": "val2"},
+            tags={"key0": "0", "key1": "val1", "key2": "val2"},
```
```suggestion
            tags={"key0": 0, "key1": "val1", "key2": "val2"},
```

If it worked before this test should probably still pass without needing this change.

ArshiaZr committed:
refactor: remove redundant checks, lift OTel length restriction, and apply minor refinements

ArshiaZr committed:
refactor: remove redundant checks, lift OTel length restriction, and apply minor refinements

ferruzzi commented on airflow/metrics/validators.py:
```diff
@@ -155,26 +155,11 @@ def stat_name_otel_handler(
-    if len(proposed_stat_name) > OTEL_NAME_MAX_LENGTH:
-        # If the name is in the exceptions list, do not fail it for being too long.
-        # It may still be deemed invalid for other reasons below.
-        for exemption in BACK_COMPAT_METRIC_NAMES:
```
My IDE is currently borked so I can't double check this easily; is BACK_COMPAT_METRIC_NAMES used anywhere else?  It's possible that can also be pruned out now if it was only ever used as a work-around for the name length bug.

ferruzzi replied:
Looking at this now, I am pretty sure this was correct.  It looks like it was only ever used there and in the related tests.   Please double check then kill it if that's right.

ArshiaZr committed:
Removed BACK_COMPAT_METRIC_NAMES as it's not beeing used anymore

ArshiaZr committed:
Removed BACK_COMPAT_METRIC_NAMES as it's not beeing used anymore

ferruzzi commented on airflow/metrics/validators.py:
```diff
@@ -157,26 +155,11 @@ def stat_name_otel_handler(
     # need the return value but will make use of the validation checks. If
     # no exception is thrown, then the proposed name meets OTel requirements.
-    stat_name_default_handler(proposed_stat_name, max_length=999 if name_length_exemption else max_length)
+    stat_name_default_handler(proposed_stat_name, max_length=999)
```
Shouldn't this now be capped at 256?

ArshiaZr replied:
yeah that's true. My bad

ferruzzi commented on tests/core/test_otel_logger.py:
```diff
@@ -142,6 +131,18 @@ def test_incr_with_rate_limit_works(self, mock_random, name):
         assert mock_random.call_count == 2
         assert self.map[full_name(name)].add.call_count == 1
 
```
Nice addition, thanks for these

ferruzzi commented on tests/core/test_otel_logger.py:
```diff
@@ -320,3 +333,43 @@ def test_timer_start_and_stop_manually_send_true(self, mock_time, name):
             name=full_name(name), callbacks=ANY
         )
+
+    def test_get_name_invalid_cases(self):
```
Non-blocking nit-pick, feel free to ignore:  All of these expected-to-fail edge cases could be one parameterized test which parameterizes the name and the expected error message.

ArshiaZr committed:
Capped otel metric name to 256 chars. Made test cases for otel_logger more consistant

ArshiaZr committed:
Capped otel metric name to 256 chars. Made test cases for otel_logger more consistant

ArshiaZr committed:
Merge remote-tracking branch 'upstream/main'

ferruzzi reviewed (commented):
Love that new parameterized test.  Thanks for doing that, you didn't have to.

Approved pending CI passing.

potiuk committed:
Fix failing ruff check on main (#44399)

potiuk committed:
Fix formatting for backport failure message. (#44392)

potiuk committed:
Remove Pydantic 2.10.0/2.10.1 workaround (#44400)

The Pydantic 2.10.0/2.10.1 workaround implemented in #44317 is not
needed any more as it has been fixed in Pydantic 2.10.2 in
the https://github.com/pydantic/pydantic/pull/10962

This PR removes the workaround and bumps min Pydantic version to
2.10.2.

perry2of5 committed:
Remove reference to hatch and add uv usage (#44401)

rawwar committed:
AIP-84 Migrate Trigger Dag Run endpoint to FastAPI  (#43875)

* init

* wip

* remove logical_date

* fix trigger dag_run

* tests WIP

* working tests

* remove logical_date from post body

* remove logical_date from tests

* fix

* include return type

* fix conf

* feedback

* fix tests

* Update tests/api_fastapi/core_api/routes/public/test_dag_run.py

* feedback

uranusjr committed:
Also check NEW_SESSION imports in pre-commit (#44388)

mobuchowski committed:
add basic system tests for OpenLineage (#43643)

Signed-off-by: Maciej Obuchowski <obuchowski.maciej@gmail.com>

uranusjr committed:
Add DAG arguments to asset decorator (#44384)

I intentionally omitted some arguments e.g. start_date. We can figure
out what to do with those later. I think these arguments are mostly
considered reasonable and not something we wish we didn't add?

melotik committed:
add openblock labs to INTHEWILD (#44365)

uranusjr committed:
Add 'airflow assets list' (#44387)

ephraimbuddy committed:
Delete `dag_version` table when downgrading the DB from the versioned Airflow (#44397)

In dagcode, fileloc_hash column was a primary key on the version before dag
versioning. If dagcode is not deleted before downgrading, the fileloc_hash column,
created when downgrading, would be a null column that can't be used for a primary key.
A similar thing applies to serialized dag, which would have duplicates, and the
primary key can't be determined. The solution is to delete the dag_version table data,
which will delete the serdag and dagcode table data

ephraimbuddy committed:
Allow 'dag_version' table to be deleted by db clean command (#44389)

Deleting a dag would cascade delete dag_version table along with
serialized_dag model and dagcode. However, we should be able to
delete dag_version directly. I didn't add dag_code or serdag model
because it won't make sense to have an existing dag_version without
a corresponding dagcode and serdag

Lee-W committed:
docs(newsfragment): add template for significant newsfragments (#44378)

* docs(newsfragment): add template for significant newsfragments

Closes: #44374

* docs(contributing-docs): add sigificant file template description

* ci(github-actions): add script to check the news fragment content

jason810496 committed:
AIP-84 Refactor SortParm (#44345)

* Refactor get_connections

* Allow Column type for `to_replace` parameter

* Refactor get_dags

* Refactor get_import_errors

* Refactor SortParam, get_dag_runs

* Fix default ordering when directly using SortParam

- related: https://github.com/apache/airflow/pull/44393

* Fix get_list_dag_runs_batch

kunaljubce committed:
Move external task sensor to standard provider (#44288)

ashb committed:
Add note about why we still have `black` config settings (#44415)

We mostly use ruff everywhere, but there is one use-case where black, or specifically blacken-docs is still needed.

kacpermuda committed:
chore: unify handling of gcs paths (#44410)

Signed-off-by: Kacper Muda <mudakacper@gmail.com>

kaxil committed:
AIP-72: Add support for fetching XComs in Supervisor (#44408)

kaxil committed:
AIP-72: Improve Supervisor and Task Instance State Validation (#44405)

gopidesupavan committed:
Move OpenAPI tests to breeze container (#44326)

* move openapi tests to breeze container

* remove space

* move openapi tests inside special tests workflow

* merge openapi tests into special tests workflow

* use postgres backend for openapi tests

* ignore tmp folder tests discovery in openapi tests

* adding retries to python client tests connection

* adding retries to python client tests connection

* export auth backend configs

* use connectivity check to verify webserver started or not

* rename is-special-tests-require to special-tests-required

* rename openapi-tests to python-api-client-tests

* rename static-checks-mypy-docs.yml to ci-image-checks.yml

* rename static-checks-mypy-docs.yml to ci-image-checks.yml

* remove special-tests-required param

eladkal committed:
Update providers metadata 2024-11-27 (#44418)

kacpermuda committed:
feat: add OpenLineage support for transfer operators between gcs and local (#44417)

Signed-off-by: Kacper Muda <mudakacper@gmail.com>

jason810496 committed:
AIP-84 Refactor test cases with `datetime_zulu_format` (#44323)

* Refactor test_dag_run

* Refactor test_assets

* Rename datetime_zulu_format to from_datetime_to_zulu

* Fix TestListDagRunsBatch

amoghrajesh committed:
Fixing rendering of code block for newsfragment command (#44419)

ArshiaZr committed:
Merge branch 'main' into main

ArshiaZr committed:
Merge branch 'main' of https://github.com/ArshiaZr/airflow

shahar1 commented on airflow/metrics/base_stats_logger.py:
```diff
@@ -31,55 +33,53 @@ class StatsLogger(Protocol):
     def incr(
         cls,
-        stat: str,
+        metric_name: str,
```
@ferruzzi / @potiuk - should we consider it as a breaking change?

ferruzzi replied:
I guess maybe we should, but it's also an internal/private method (not part of the public API AFAICT)  so I don't know if we really need to.

ferruzzi replied:
I've asked Arshia to revert that part of the change.  I like the idea, the name is way more clear, but we can deal with that discussion another time.  I'd like to get this merged as it is holding up a little project that he and Danny are working on.

shahar1 reviewed (commented):
Looks good overall, I've got a couple of questions

shahar1 committed:
Merge branch 'main' into main
